pretyping/typing.ml

```diff
@@ -49,6 +49,30 @@ let e_assumption_of_judgment env evdref j =
   with Type_errors.TypeError _ | PretypeError _ ->
     error_assumption env !evdref j
 
+let e_judge_of_applied_inductive_knowing_parameters env evdref funj ind argjv =
+  let rec apply_rec n typ = function
+  | [] ->
+      { uj_val  = mkApp (j_val funj, Array.map j_val argjv);
+        uj_type =
+          let ar = inductive_type_knowing_parameters env !evdref ind argjv in
+          hnf_prod_appvect env !evdref (EConstr.of_constr ar) (Array.map j_val argjv) }
+  | hj::restjl ->
+      match EConstr.kind !evdref (whd_all env !evdref typ) with
+      | Prod (_,c1,c2) ->
+         if Evarconv.e_cumul env evdref hj.uj_type c1 then
+           apply_rec (n+1) (subst1 hj.uj_val c2) restjl
+         else
+           error_cant_apply_bad_type env !evdref (n, c1, hj.uj_type) funj argjv
+      | Evar ev ->
+          let (evd',t) = Evardefine.define_evar_as_product !evdref ev in
+          evdref := evd';
+          let (_,_,c2) = destProd evd' t in
+          apply_rec (n+1) (subst1 hj.uj_val c2) restjl
```

Comment:
Why don't we unify the domain with the type of the argument here and in e_judge_of_apply?

Reply:
Well, that's a very good question. Seems like we shall (in theory) loose solutions otherwise. I don't know how to build an example which shall reach this line in `typing.ml` though.

I shall do a different PR for it.